### ISSUE TITLE
glideN64 is corrupted for most games

### ISSUE THREAD

**Issue by Panderner** [gonetz/GLideN64#2089]
Hi the graphics are almost corrupted most games even glideN64-Very-Accurate on mupen64plus fz and libretro, I'm using PowerVR ge8320 (mediatek Helio p22) android phone

Comparison:

Libretro mupen64plus-next (normal and gles3 versions):
![Screenshot_2019-09-23-13-39-43-97_a845f06d2c9842610668be6995f9e34f](https://user-images.githubusercontent.com/37503397/65403924-ce138080-de08-11e9-8171-f36139b3545f.png)
![Screenshot_2019-09-23-13-43-05-21_a845f06d2c9842610668be6995f9e34f](https://user-images.githubusercontent.com/37503397/65404464-33687100-de0b-11e9-8f75-aa2949ef0ec3.png)

Mupen64plus fz:
In glideN64 very accurate the polygon and textures are glitchy
![ezgif com-video-to-gif](https://user-images.githubusercontent.com/37503397/65404379-cc4abc80-de0a-11e9-9eb9-9a6023f9a93e.gif)

In glideN64 medium same as libretro core

Super Mario 64 used for comparison and it happened for most games

**Comment by m4xw** [comment]
These Issues started appearing when pulling the following changes in from upstream: https://github.com/libretro/mupen64plus-libretro-nx/commit/16baae76c75e434d161b4974c1bcc1cca6db9657

**Comment by gonetz** [comment]
@m4xw the commit contains almost all changes since the PR 4.0
If we want to find the problem, we need to bisect to the wrong commit in GLideN64 repo.
@fzurita can you find, when the problem appeared?

**Comment by gonetz** [comment]
@fzurita I just tested Mupen64plus fz from Play Market. It works correct. Which version of GLideN64 it uses?

**Comment by m4xw** [comment]
@gonetz he told me 2 days ago that he didn't update it yet, so I assume the reporter compiled it himself.

**Comment by m4xw** [comment]
However we suspect it's related to the shader changes.
fwiw shader cache breaks too on nvidia shield android with gles, no issues with OpenGL (on switch via nouveau, and lakka switch with nvidia drivers)

**Comment by gonetz** [comment]
@m4xw  I modified shaders in commit 2d712f2ae048.
Could you make android build from previous commit 993960f9e and compare with 2d712f2ae048?

**Comment by m4xw** [comment]
@gonetz Sorry but I dont have a dev env set up since we switched NDK version.
I was refering to the exact shader modifications you did, but it's not easy to bisect these issues with my fork since I always have to rebase all my commits and on top the commit you showed, also would be broken anyway without a cherrypick of 34c6cae36485d06d050e7eaea7d4092d42f47e23
Can do it at my workstation at the office but not before next week :/

**Comment by fzurita** [comment]
I'll go ahead and take a look, I have some time today. @Panderner which phone do you have?

**Comment by m4xw** [comment]
Thanks man, that would be great.
@fzurita 
> PowerVR ge8320 (mediatek Helio p22)

You will also find some more infos on my fork's issue tracker https://github.com/libretro/mupen64plus-libretro-nx/issues/99

This might also be related https://github.com/libretro/mupen64plus-libretro-nx/issues/101
And it might've also caused a regression for bgStriped on Windows+GL https://github.com/libretro/mupen64plus-libretro-nx/issues/98
Weirdly enough that seems linked to monitor refresh rate.
I didn't confirm Yoshi's pacing issues, but I have a report that this also appeared when pulling in all the new changes from the commit i mentioned.
*runs away*

**Comment by fzurita** [comment]
So in two devices so far, I can't reproduce the issue, they are Mali and Adreno devices. I do have a PowerVR device to try as well. Hopefully I see it there.

**Comment by m4xw** [comment]
Some Mali devices seem affected tho https://github.com/libretro/mupen64plus-libretro-nx/issues/99#issuecomment-534497518

>I have to rebuild a RPi2/3 image will then check & report again. But since I had these glitches on a Khadas VIM, VIM 3 & RockPro64 which use a Mali 450, T860 & G52 I suspect it's related to them.

Fwiw it was reported with no issues on RPi
>Just tested on RPI3B+ with KMS/DRM/GLES2 using latest develop branch and all default core options, all looks normal here.

**Comment by m4xw** [comment]
Also some amlogic devices.
Here was another Mali:
> I used an Android TV Box, MX3-G S912, Open GL 2.0ES and Mali 450.

Broke like this:
![grafik](https://user-images.githubusercontent.com/13141469/65819903-7028d380-e222-11e9-9236-9e3ba8c5b769.png)

**Comment by m4xw** [comment]
@fzurita You are testing latest master right now?
Or the commit gonetz asked?

**Comment by fzurita** [comment]
Latest master

**Comment by m4xw** [comment]
Dunno if he changed smth but when I asked people to test my experimental branch (which was what gonetz just rebased), same issues appeared..
hm...

Try this commit https://github.com/gonetz/GLideN64/commit/2d712f2ae0482d7da3b49397902551addd676107
And cherry pick https://github.com/gonetz/GLideN64/commit/34c6cae36485d06d050e7eaea7d4092d42f47e23

**Comment by m4xw** [comment]
btw are you testing GLES2 or GLES3?

**Comment by fzurita** [comment]
GLES3, I have some devices with GLES2.0 to test with, so I'll try those next.

**Comment by m4xw** [comment]
Someone said GLES3 somewhat fixed it on his Mali, others reported no benefit.
Might also be related to a few option combos.

**Comment by m4xw** [comment]
A quick reminder how it ended up for amlogic https://www.youtube.com/watch?v=6oYoRE0gkFw

**Comment by GranvilleR** [comment]
In the Libretro core, the first time I booted a game it worked correctly. I had to close and reload the content to trigger the most serious glitches (like the missing graphics in Mario 64 seen in the first post). I haven't tested the issue in Mupen FZ to know how to trigger it there. Using a Galaxy S10 Adreno.

**Comment by m4xw** [comment]
Yea that was the shadercache issue.

**Comment by fzurita** [comment]
Was the shader cache issue a libretro only issue?

**Comment by m4xw** [comment]
I can't tell you that.
However we let GLN64 handle that completely, don't have any logic there that diverges from upstream.

**Comment by m4xw** [comment]
Doesn't happen on mupen fz?

**Comment by m4xw** [comment]
@fzurita btw my rebase was done on this directly:
![grafik](https://user-images.githubusercontent.com/13141469/65820326-be8ca100-e227-11e9-9e5a-25480f4b8957.png)
I'd really test that commit to exclude a later fix if you have trouble reproducing it.
Also make sure to test the shader cache, I don't think that would be libretro exclusive and might be the actual cause of these severe bugs.

**Comment by fzurita** [comment]
I'm not seeing the issue in mupen64plus FZ with at least 3 different phones.

**Comment by m4xw** [comment]
On that commit?

**Comment by fzurita** [comment]
Latest master.

**Comment by m4xw** [comment]
Try that commit please, so we can exclude a later fix.
If we know it happened there and not later, would help a lot already.

**Comment by fzurita** [comment]
Also, the shader cache issue should be only GLES 3.0 since GLES 2.0 doesn't support the shader cache. So, there seem to be several distinct issues.

**Comment by fzurita** [comment]
I'll try that commit.

**Comment by m4xw** [comment]
Thanks.

> GLES 2.0 doesn't support the shader cache

Good to know.

**Comment by m4xw** [comment]
(Also i should probably hide the shader cache option for GLES2 then, 🙈 )

**Comment by m4xw** [comment]
Btw, which games are you testing?

**Comment by fzurita** [comment]
Well, to be more specific, the binary shader cache doesn't work (the .shaders file). The .key files I think still do work, but all shaders have to be recompiled at startup.

**Comment by m4xw** [comment]
@GranvilleR remember if it was GLES2 or 3 that you tested?

**Comment by m4xw** [comment]
also @fzurita might be worth testing this issue without savestates (in case you do use them).

**Comment by GranvilleR** [comment]
@m4xw It was GLES3.

**Comment by drinfernoo** [comment]
@fzurita Can confirm I'm not seeing really any of these bugs in FZ.

**Comment by m4xw** [comment]
@drinfernoo Its not about the playstore version.

**Comment by fzurita** [comment]
Try this build of mupen64plus fz with the latest GLideN64:

https://drive.google.com/open?id=1SJMuccm3Ooob4BicmMVAd87qlUZqC4f9

**Comment by drinfernoo** [comment]
@fzurita I can't get this .apk to install on either my phone or Shield. I get "App not installed."

**Comment by fzurita** [comment]
Whoops sorry about that. I'll make a new build, apparently I made an incremental build.

**Comment by m4xw** [comment]
I get the corrupted :/

**Comment by fzurita** [comment]
Try this one: https://drive.google.com/open?id=1kPug87NBBAK2Dvp1lvnzipC8XCf_oOSf

**Comment by m4xw** [comment]
@fzurita did you supply him GLES2 or 3?

**Comment by fzurita** [comment]
It's using EGL, so it will adapt to whatever device he has.

I do have some good new though, I'm able to reproduce the Mario Kart issues in the ADT-2 device.

**Comment by m4xw** [comment]
Lol we force it compiletime, I guess if they use GLES2 builds on GLES3 hw it would be undefined behaviour on top?

> I do have some goods new though, I'm able to reproduce the Mario Kart issues in the ADT-2 device.

You got me sweating already :P

**Comment by m4xw** [comment]
fwiw forcing gles2 when gles3 is provided fixed issues with other emu's I had before, namely ppsspp.

**Comment by fzurita** [comment]
With Android, you can't really force GLES 2.0 in a GLES 3.0 device, it will always use the latest. What you can do is limit yourself to a certain subset of the GLES API.

**Comment by m4xw** [comment]
Wanna try this hack and see if u can repro?
https://github.com/libretro/mupen64plus-libretro-nx/blob/develop/GLideN64/src/Graphics/OpenGLContext/opengl_GLInfo.cpp#L28-L32

**Comment by m4xw** [comment]
But i guess if u have a indication already, so w/e

**Comment by fzurita** [comment]
Maybe if I get stuck I'll try that.

**Comment by m4xw** [comment]
He is testing on a shield, so it should be GLES3.
However he had the same issues tho for retroarch, but I dunno which core ver he actually used (GLES3 with forced GLES2 / GLES3).

**Comment by fzurita** [comment]
And the problem is.... Shader compilation error, shouldn't be hard to figure out.

**Comment by m4xw** [comment]
heh, yea.
Wanna quote the error?

**Comment by fzurita** [comment]
Nvidia shield supports full OpenGL in Android, so Mupen64plus FZ will use that. Although, NVidia shield can support switching between GLES and Full open GL.

**Comment by m4xw** [comment]
Oh yea, I didn't know you supported GL there too, we don't do that for Android builds (yet).

**Comment by fzurita** [comment]
So the ADT-2 drivers kind of suck, so it's not giving me what line number in the shader failed.... But at least I know which shader is failing.

**Comment by fzurita** [comment]
Weird, even the vertex shader is failing:
```
2019-09-28 17:35:25.549 6523-6562/paulscode.android.mupen64plusae.GameActivity E/GLideN64: glsl_Utils.cpp:65,ERROR, "Error in vertex shader"
2019-09-28 17:35:25.549 6523-6562/paulscode.android.mupen64plusae.GameActivity E/GLideN64: glsl_Utils.cpp:73,ERROR, "#version 100			
    #if (__VERSION__ > 120)	
    # define IN in				
    # define OUT out			
    #else						
    # define IN attribute		
    # define OUT varying		
    #endif // __VERSION		
    IN highp vec4 aRectPosition;						
    													
    OUT lowp vec4 vShadeColor;							
    uniform lowp vec4 uRectColor;						
    void main()										
    {													
      gl_Position = aRectPosition;						
      vShadeColor = uRectColor;						
    } 
    "
```

**Comment by m4xw** [comment]
version 100? lol
detection failing?

**Comment by fzurita** [comment]
Yeah, that's a good point....

**Comment by fzurita** [comment]
So, it's hard coded to use ```version #100`` for all shaders in GLES2.

**Comment by m4xw** [comment]
Hm what is expected for GLES2?
200?

**Comment by fzurita** [comment]
Well, version 100 has been working fine for years. Probably something recent broke it, like using some feature of GLSL that is not supported in version 100.

**Comment by fzurita** [comment]
Probably this broke it: 2d712f2ae0482d7da3b49397902551addd676107

**Comment by m4xw** [comment]
The fz apk worked fine for him btw

**Comment by fzurita** [comment]
Probably because it's not using GLES 2.0.

**Comment by m4xw** [comment]
Yep, just wanted to confirm tho since he didnt post here.

**Comment by drinfernoo** [comment]
@fzurita @m4xw Yeah, sorry... talking in Discord as well. I haven't had any issues with game accuracy in FZ before, or with the newer .apk you sent a bit ago.

**Comment by fzurita** [comment]
Ok, reverted back to commit https://github.com/gonetz/GLideN64/commit/2d712f2ae0482d7da3b49397902551addd676107 and everything still works there... Must be something more recent.

**Comment by fzurita** [comment]
My mistake https://github.com/gonetz/GLideN64/commit/2d712f2ae0482d7da3b49397902551addd676107 is indeed the broken one. I had a compilation error and I was still running the old version without realizing it.

**Comment by m4xw** [comment]
Yea you need to cherry pick the compile fix

**Comment by m4xw** [comment]
Did you try 1 earlier yet?

**Comment by fzurita** [comment]
Yes, previous revision had no words.

**Comment by fzurita** [comment]
I do have a line number actually, somehow I missed it earlier:

```
2019-09-28 18:47:13.457 8629-8679/paulscode.android.mupen64plusae.GameActivity E/GLideN64: glsl_Utils.cpp:40,ERROR, "shader_compile error: 0:20: L0002: Undeclared variable 'gl_FragDepth'
    "
2019-09-28 18:47:13.457 8629-8679/paulscode.android.mupen64plusae.GameActivity E/GLideN64: glsl_Utils.cpp:65,ERROR, "Error in vertex shader"
2019-09-28 18:47:13.457 8629-8679/paulscode.android.mupen64plusae.GameActivity E/GLideN64: glsl_Utils.cpp:73,ERROR, "#version 100 
    #extension GL_EXT_shader_texture_lod : enable 
    #extension GL_OES_standard_derivatives : enable 
    #if (__VERSION__ > 120)		
    # define IN in					
    # define OUT out				
    # define texture2D texture		
    #else							
    # define IN varying			
    # define OUT					
    #endif // __VERSION __			
    IN mediump vec2 vTexCoord0;							
    uniform sampler2D uTex0;								
    uniform sampler2D uTex1;								
    OUT lowp vec4 fragColor;								
    														
    void main()											
    {														
    	fragColor = texture2D(uTex0, vTexCoord0);			
    	gl_FragDepth = texture2D(uTex1, vTexCoord0).r;	
      gl_FragColor = fragColor; 
    } 
    
    "
```

**Comment by fzurita** [comment]
So fixing that error doesn't fix anything.

Not sure what is wrong a this point, the generated shaders are almost identical to before the change. Mario Kart 64 doesn't even use the texture modes that would make the changes take effect, so the shaders generated are pretty much identical.

The only real difference is that we got rid of the "realWidth" and "realHeight", so it's possible that GLES 2.0 is not happy with the new texture heights and widths.

**Comment by Panderner** [comment]
Nothing happens on latest build

**Comment by Panderner** [comment]
I'm using realme c2

**Comment by gonetz** [comment]
`gl_FragDepth = texture2D(uTex1, vTexCoord0).r;
`
I guess GLES2 does not support gl_FragDepth modification.

**Comment by gonetz** [comment]
>The only real difference is that we got rid of the "realWidth" and "realHeight", so it's possible that GLES 2.0 is not happy with the new texture heights and widths.

Hardly. "realWidth" and "realHeight" are atavism from old days, when hardware did not support non-pot texture sizes. "width" was for N64 texture width, "realWidth" was for power-of-two width.
I removed conversion to power-of-two texture sizes long ago, but "realWidth" and "realHeight"  remained and most of time they duplicated "width" and "height". "realWidth" and "width" were different in rare cases, when N64 texture mirroring emulated via creating mirrored texture, that is when texture contained original and mirrored image.

**Comment by fzurita** [comment]
Gotcha, I'm running out of ideas as to what is different that would be different. Generated shaders are the same as before. Texture widths and heights should be the same as well. I don't see much else different.

**Comment by m4xw** [comment]
I saw this before, maybe related to this hack?
Running out of ideas too..
![grafik](https://user-images.githubusercontent.com/13141469/65831865-2261a900-e2be-11e9-978a-5ed79c60adfd.png)

**Comment by fzurita** [comment]
It's not, I already tried removing it just in case, it had no effect.

**Comment by fzurita** [comment]
Ok, I think I got it narrowed down, at least in Mario Kart. Replacing this code:
```
		m_part =
			"highp vec2 clampWrapMirror(in highp vec2 vTexCoord, in highp vec4 vClamp,		\n"
			"	in highp vec2 vWrap, in lowp vec2 vMirror, in highp vec2 vScale)			\n"
			"{																				\n"
			"  highp vec2 texCoord = clamp(vTexCoord, vClamp.xy, vClamp.zw);				\n"
			"  lowp vec2 one = vec2(1.0);													\n"
			"  lowp vec2 clamped = step(vClamp.zw, texCoord);								\n"
			"  lowp vec2 notClamped = one - clamped;										\n"
			"  texCoord = clamped * texCoord + notClamped * mod(texCoord, vWrap);			\n"
			"  highp vec2 intPart = floor(texCoord);															\n"
			"  highp vec2 fractPart = fract(texCoord);								\n"
			"  lowp vec2 needMirror = step(vec2(0.5), mod(intPart, vec2(2.0))) * vMirror;	\n"
			"  texCoord = clamped * texCoord + notClamped * fractPart;						\n"
			"  texCoord = (one - vMirror) * texCoord + vMirror * fractPart;					\n"
			"  texCoord = (one - texCoord) * needMirror + texCoord * (one - needMirror);	\n"
			"  texCoord *= vScale;															\n"
			"  return texCoord;																\n"
			"}																				\n"
			;
```

with:
```
		m_part =
				"highp vec2 clampWrapMirror(in highp vec2 vTexCoord, in highp vec4 vClamp,		\n"
				"	in highp vec2 vWrap, in lowp vec2 vMirror, in highp vec2 vScale)			\n"
				"{																				\n"
				"  return vTexCoord;																\n"
				"}																				\n"
				;
```

and everything works.

**Comment by fzurita** [comment]
Weird, now that I started messing with it, I'm getting this error:

```
"shader_compile error: 0:29: P0004: High precision not supported, instead compiling high precision as medium precision
```

**Comment by fzurita** [comment]
Well... getting closer. Floating point math is failing due to medium precision. Changing this:
```
		std::array<f32, 4> aTexClamp[2] = { { -10000.0f, -10000.0f, 10000.0f, 10000.0f },
											{ -10000.0f, -10000.0f, 10000.0f, 10000.0f } };
		std::array<f32, 2> aTexWrap[2] = { { 10000.0f, 10000.0f }, { 10000.0f, 10000.0f } };
```

to this fixes it partially:
```
		std::array<f32, 4> aTexClamp[2] = { { -1000.0f, -1000.0f, 1000.0f, 1000.0f },
											{ -1000.0f, -1000.0f, 1000.0f, 1000.0f } };
		std::array<f32, 2> aTexWrap[2] = { { 1000.0f, 1000.0f }, { 1000.0f, 1000.0f } };
```

**Comment by fzurita** [comment]
Pull request has been made: https://github.com/gonetz/GLideN64/pull/2092

**Comment by gonetz** [comment]
@Panderner fix is done. Please test again.

**Comment by drinfernoo** [comment]
I've tested builds of the core from @m4xw, and I'm not seeing any improvement on my end.

**Comment by m4xw** [comment]
@drinfernoo You issue might be unrelated.

**Comment by m4xw** [comment]
Shanti quickly confirmed the fix for MK64 from remote, he will do more throughout testing later.
So yea that GLES2 issue is very likely fixed.
Still no idea what ghosts hunt drinfernoo's shield.

**Comment by fzurita** [comment]
@drinfernoo it sounds like you have a different issue. Do you have a Nvidia shield tv? Is your problem only work the libretro core?

**Comment by Panderner** [comment]
@gonetz where is the build?

**Comment by drinfernoo** [comment]
@fzurita Yes and yes. FZ works fine.

**Comment by fzurita** [comment]
Ok, I have a Nvidia shield tv. I'll check what is happening and whether it's related to GLideN64.

**Comment by gonetz** [comment]
@gonetz where is the build?

Where did you get it for the first time?
Sources updated, so binaries should be updated too.

**Comment by fzurita** [comment]
Shader cache definitely breaks the shield tv with the libretro core. It works fine on mupen64plus fz. The problem is probably in the libretro core and not the video plug-in.

This issue should probably close.

**Comment by gonetz** [comment]
As I understand, we did everything we could do on our side.
Lets close it then.

**Comment by m4xw** [comment]
Ye the libretro issue is only semi related, basically your changes broke some of my GL state machine logic (which should never happen to begin with).
But my glsm version does some pretty funky stuff to begin with, just wasn't a issue before your recent updates.
So, yea. It's my homework not yours ;)

**Comment by Panderner** [comment]
@gonetz glideN64 still not fixed and updated

**Comment by gonetz** [comment]
@Panderner fzurita says that GLideN64 works fine on mupen64plus fz.
Can we do anything on our side to make it working better on Libretro mupen64plus?

**Comment by m4xw** [comment]
@gonetz No need to do anything from your part, in regards to libretro.
I first need to understand what actually triggered the Issue in my codebase.
fwiw some issues only seem to appear at lowest resolution only.

**Comment by Panderner** [comment]
Still not fixed I'm using realme C2, the mupen64plus fz and libretro mupen64plus didn't update glideN64 and still corrupted

**Comment by fzurita** [comment]
@Panderner So does it work with Mupen64Plus FZ for you but not with libretro?

**Comment by Panderner** [comment]
Even gliden64 fast, medium, accurate, and very accurate, glitched out

**Comment by fzurita** [comment]
This is a completely different problem between libretro and mupen64plus fz. Please open a new issue that only describes the problem in mupen64plus fz.

**Comment by m4xw** [comment]
@fzurita We don't have any of these configurations, he must be talking about fz.

**Comment by m4xw** [comment]
@gonetz Just a headsup, this issue isn't fixed, I found the cause and it does indeed affect mupen fz and mupen next, only at lowest resolution (< 640x580).
I default to 320x240, hence I got so many reports and was still never able to reproduce it, heh.
tldr of it is: You mixed some logic realH/W and width/height do and your changes break your own design completely.
You used to set realWidth etc, do the buffer size calc and then you overwrote it later, now it's all the same and it tries to blit a 640x580 texture onto a 320x240 fb (instead of x298 anyway, due to just picking DisplayWindow sizes).
Dunno if you want to deprecate it or it was a oversight.
This is likely also related to #2102, as well:
https://github.com/libretro/mupen64plus-libretro-nx/issues/109
https://github.com/libretro/mupen64plus-libretro-nx/issues/108
https://github.com/libretro/mupen64plus-libretro-nx/issues/98
I will write a Issue about that & the details sometime tomorrow.

**Comment by m4xw** [comment]
However I believe the specific issue Panderner had, is something completely different tho.

**Comment by fzurita** [comment]
Yeah, I created this issue for Panderner's issues: https://github.com/gonetz/GLideN64/issues/2108

**Comment by gonetz** [comment]
It is very strange decision to set default to 320x240 for N64 games, because half of them use higher internal resolution. The plugin designed to work in high or at least native resolutions, not in resolutions below native.
The fzurita's suggestion to set nativeResFactor to 1 should work indeed. I tested hi-res N64 games such as Indi in 320x240 resolution with native-res enabled and found no issues. What "scaled now ofc" means?

**Comment by m4xw** [comment]
@gonetz This is a bit of a unexpected response, since it used to work before your changes.
> The plugin designed to work in high or at least native resolutions, not in resolutions below native.

So I must assume it was unknowingly working?

320x (no nativeResFactor, no longer working)
![image](https://user-images.githubusercontent.com/13141469/67376535-996c2380-f584-11e9-8564-90f7abc8ed48.png)

320x (with nativeResFactor)
![image](https://user-images.githubusercontent.com/13141469/67376581-ae48b700-f584-11e9-9cff-80f686b8dbb1.png)

With scaling I mean the scale factor calculations that are applied in your blitting logic, but I guess, the pics tell it already.

Also thats one of the cases where you broke the logic that worked prior:
![image](https://user-images.githubusercontent.com/13141469/67376874-1dbea680-f585-11e9-8986-5efda21c2e5b.png)

Notice the ->width etc override, I confirmed that with my setup it would override them to 320x298
So, what does GLideN want going forward?

**Comment by gonetz** [comment]
> So I must assume it was unknowingly working?

I did not know that it might work that way. It is really wrong idea to have rendering buffer smaller than frame buffer in RDRAM.

> Also thats one of the cases where you broke the logic that worked prior:

It was major refactoring, and I really could made mistakes. This change looks as a mistake.
However, I can't see how it is related to the original issue. Could you open a new issue with whole details about what is wrong and when (and why) things got broken?

> So, what does GLideN want going forward?

I'm not sure that I understand that question.

**Comment by m4xw** [comment]
(just a quick reminder i didnt forget to create the issue, but I didn't have time to investigate some more, for creating a detailed bug report).
Hope I get to it on the holidays.
